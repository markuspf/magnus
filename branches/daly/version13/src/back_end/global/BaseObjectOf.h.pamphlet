\documentclass{article}
\usepackage{axiom}
\begin{document}
\title{src/backend/global/BaseObjectOf.h}
\author{Stephane Collart}
\maketitle
\begin{abstract}
\end{abstract}
\eject
\tableofcontents
\eject
\section{include}
<<include>>=
#include "PureRep.h"
#include "ObjectOf.h"

@
\section{template<class Rep> class BaseObjectOf}
<<template<class Rep> class BaseObjectOf>>=
template<class Rep> class BaseObjectOf : private ObjectOf<Rep>
{
// This assumes that Rep is derived from PureRep
public:
  // copy constructor provided by compiler
  // destructor provided by compiler
  // assignment operator provided by compiler

protected:
  // Inherited from base; adjust access:
  ObjectOf<Rep>::look;
  ObjectOf<Rep>::change;
  ObjectOf<Rep>::enhance;
  ObjectOf<Rep>::acquireRep;
  BaseObjectOf( Rep* newrep ) : ObjectOf<Rep>( newrep) { }

private:
	// Dummy Forcing Member:
	void force_derivation( ) { const PureRep* rc = look(); }
	// With this member PureRep is forced to be 
        // an accessible base class of Rep
};

@
\section{license}
<<license>>=
// Copyright (C) 1994-2004 The New York Group Theory Cooperative
// See license/NYGTC for the full notice
@
<<*>>=
// src/backend/global/BaseObjectOf.h
#ifndef _BASE_OBJECT_OF_H_
#define _BASE_OBJECT_OF_H_

<<license>>
<<include>>
<<template<class Rep> class BaseObjectOf>>
#endif

@
\eject
\begin{thebibliography}{99}
\bibitem{1} nothing
\end{thebibliography}
\end{document}
