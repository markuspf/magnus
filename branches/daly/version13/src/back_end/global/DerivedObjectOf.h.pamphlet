\documentclass{article}
\usepackage{axiom}
\begin{document}
\title{src/backend/global/DerivedObjectOf.h}
\author{Stephane Collart}
\maketitle
\begin{abstract}
\end{abstract}
\eject
\tableofcontents
\eject
\section{class DerivedObjectOf>
<<class DerivedObjectOf>>=
template <class Base, class Rep> class DerivedObjectOf : public Base
{

// This class assumes Base is itself an ObjectOf kind of class, i.e.
// 1) Base has members look(), enhance(), change() returning some BaseRep* type
// 2) Base has a constructor Base( BaseRep* );
// 3) BaseRep is directly or indirectly derived from PureRep
// 4) Rep is derived from BaseRep

public:
	// copy constructor provided by compiler
	// destructor provided by compiler
	// assignment operator provided by compiler

protected:
	// Shadowed representation accessors:
	const Rep* look( ) const { return (Rep*)Base::look(); }
	Rep* enhance( ) const { return (Rep*)Base::enhance(); }
	Rep* change( ) { return (Rep*)Base::change(); }

	DerivedObjectOf( Rep* newrep ) : Base( newrep ) { }
	// for base class initialisation

	DerivedObjectOf( const Base& b ) : Base( b ) { }
	// for base class intialisation when cast conversion
        // to derived is desired 

};

@
\section{license}
<<license>>=
// Copyright (C) 1994-2004 The New York Group Theory Cooperative
// See license/NYGTC for the full notice
@
<<*>>=
// src/backend/global/DerivedObjectOf.h
#ifndef _DERIVED_OBJECT_OF_H_
#define _DERIVED_OBJECT_OF_H_

<<license>>
<<class DerivedObjectOf>>
#endif
@
\eject
\begin{thebibliography}{99}
\bibitem{1} nothing
\end{thebibliography}
\end{document}
