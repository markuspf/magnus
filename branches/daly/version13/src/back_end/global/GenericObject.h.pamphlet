\documentclass{article}
\usepackage{axiom}
\begin{document}
\title{src/backend/global/GenericObject.h}
\author{Stephane Collart}
\maketitle
\begin{abstract}
\end{abstract}
\eject
\tableofcontents
\eject
\section{include}
<<include>>=
#include "PureRep.h"
#include "BaseObjectOf.h"
#include "Type.h"

@
\section{struct GenericRep}
a common root class for all object-level (see PureObject.h) classes
with a coherent type scheme.

See PureObject

In addition, this class provides explicit type query handles; classes
derived from this class will normally derive *publicly*, to permit
assignment by non-friend
<<struct GenericRep>>=
struct GenericRep : public PureRep
{
	// default constructor provided by compiler
	// copy constructor provided by compiler
	// assignment operator disabled

	static const Type theGenericObjectType;
	// every representation class derived from this 
        // should carry a similarly defined type field

	static Type type() { return theGenericObjectType; }
	// every representation class derived from this 
        // should carry a similarly defined type member

	virtual Type actualType() const { return type(); }
	// this member serves as the root of all type querying members in this
	// type scheme
	// every representation class derived from this should override this
	// member similarly to return the actual type of the class

	PureRep* clone( ) const { return new GenericRep(*this); }

};

@
\section{class GenericObject}
<<class GenericObject>>=
class GenericObject : protected BaseObjectOf<GenericRep>
{
public:
	GenericObject( ) : BaseObjectOf<GenericRep>( new GenericRep ) { }

	// copy constructor provided by compiler
	// destructor provided by compiler
	// both of the preceding ref-counting compatible
	// assignment operator provided by compiler
	// ref-counting compatible

	static Type type() { return GenericRep::type(); }
	// this `type querying' member is provided solely as an example, 
        // since this class is pseudo-abstract (ie. its representation 
        // class is abstract) and no actual objects of this type will 
        // ever be created

	Type actualType() const { return look()->actualType(); }

protected:
	// look(), change() and enhance() inherited

	// Special wrapping constructor to wrap new representations (returned
	// by eg. delegated methods) and for base initialisation by derived
	// classes:

	GenericObject( GenericRep* newrep ) : BaseObjectOf<GenericRep>(newrep) { }

};

@
\section{license}
<<license>>=
// Copyright (C) 1994-2004 The New York Group Theory Cooperative
// See license/NYGTC for the full notice
@
<<*>>=
// src/backend/global/GenericObject.h
#ifndef _GENERIC_OBJECT_H_
#define _GENERIC_OBJECT_H_


<<include>>
<<license>>
<<struct GenericRep>>
<<class GenericObject>>
#endif
@
\eject
\begin{thebibliography}{99}
\bibitem{1} nothing
\end{thebibliography}
\end{document}
