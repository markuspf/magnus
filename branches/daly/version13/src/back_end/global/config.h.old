// Copyright (C) 1994 The New York Group Theory Cooperative
// See magnus/doc/COPYRIGHT for the full notice.

// Contents: Configuration variables set up by the config script
//
// Principal Author: Jon Lennox
//
// Status: Useable.
//
// Revision History:
//

#ifndef _CONFIG_H_
#define _CONFIG_H_

#define MAGNUS_HOME "/usr/bormotov/magnus"
// We keep this compiled-in value to ease debugging.
// Use MagnusHome::magnusHome() in back_end/general/ for
// the Magnus root directory.

#define LINUX

#define MAGNUS_TMP "/tmp"

#endif /* _CONFIG_H_ */
