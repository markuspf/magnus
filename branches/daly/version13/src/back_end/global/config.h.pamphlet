\documentclass{article}
\usepackage{axiom}
\begin{document}
\title{src/backend/global/config.h}
\author{Jon Lennox}
\maketitle
\begin{abstract}
\end{abstract}
\eject
\tableofcontents
\eject
\section{license}
<<license>>=
// Copyright (C) 1994-2004 The New York Group Theory Cooperative
// See license/NYGTC for the full notice
@
<<*>>=
// src/backend/global/config.h
#ifndef _CONFIG_H_
#define _CONFIG_H_

<<license>>

#define MAGNUS_HOME "/usr/local/magnus"
// We keep this compiled-in value to ease debugging.
// Use MagnusHome::magnusHome() in back_end/general/ for
// the Magnus root directory.

#define LINUX

#define MAGNUS_TMP "/tmp"

#endif /* _CONFIG_H_ */
@
\eject
\begin{thebibliography}{99}
\bibitem{1} nothing
\end{thebibliography}
\end{document}
