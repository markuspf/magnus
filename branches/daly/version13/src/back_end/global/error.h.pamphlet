\documentclass{article}
\usepackage{axiom}
\begin{document}
\title{src/backend/global/error.h}
\author{Roger Needham}
\maketitle
\begin{abstract}
\end{abstract}
\eject
\tableofcontents
\eject
\section{error}
<<error>>=
#ifdef __GNUG__
//@jml This tells g++ that error does not return.
inline volatile void error(const char *msg) {
#else
inline void error(const char* msg) {
#endif

  cout << "fatal_error_ {" << msg << '}' << endl << flush;
  // Try to shove the message directly out to the front end
  cerr << endl << "Fatal error: " << msg << endl;
  while ( 1 );
  // So we can get a stack backtrace when debugging
}

@
\section{warn}
<<warn>>=
inline void warn(const char* msg) {
  cerr << endl << "Warning: " << msg << endl;
}

@
\section{license}
<<license>>=
// Copyright (C) 1994-2004 The New York Group Theory Cooperative
// See license/NYGTC for the full notice
@
<<*>>=
// src/backend/global/error.h
#ifndef _ERROR_H_
#define _ERROR_H_

<<license>>
<<error>>
<<warn>>
#endif
@
\eject
\begin{thebibliography}{99}
\bibitem{1} nothing
\end{thebibliography}
\end{document}
