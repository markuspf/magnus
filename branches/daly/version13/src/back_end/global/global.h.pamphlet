\documentclass{article}
\usepackage{axiom}
\begin{document}
\title{src/backend/global/global.h}
\author{Roger Needham}
\maketitle
\begin{abstract}
\end{abstract}
\eject
\tableofcontents
\eject
\section{include}
<<include>>=
#include <stream.h>
#include <iostream.h>
#include <fstream.h>
#include <strstream.h>
//@rn #include <string.h>
#include <ctype.h>
#include <stdlib.h>
#include <math.h>

#include "error.h"
#include "IPC.h"
#include "Trichotomy.h"

@
\section{global.h}
Contents: Global declarations to be included by all other header files.
\section{license}
<<license>>=
// Copyright (C) 1994-2004 The New York Group Theory Cooperative
// See license/NYGTC for the full notice
@
<<*>>=
// src/backend/global/global.h
#ifndef _GLOBAL_H_
#define _GLOBAL_H_

<<include>>
<<license>>

//#define NULL 0

#define SAFETY 2

#define TRUE  1
#define FALSE 0

typedef int Bool;

#define INFINITE 0


//------ global function prototypes, variables, inline functions ------------//

inline int min(int i, int j) { return ((i<j) ? i : j); }
inline int max(int i, int j) { return ((i>j) ? i : j); }
inline int abs(int i) { return ((i < 0) ? -i : i); }
inline int sign(int i) { return (i == 0) ? 0 : ( (i > 0) ? 1 : -1 ); }
#endif
@
\eject
\begin{thebibliography}{99}
\bibitem{1} nothing
\end{thebibliography}
\end{document}
