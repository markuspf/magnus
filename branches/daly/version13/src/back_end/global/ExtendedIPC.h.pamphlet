\documentclass{article}
\usepackage{axiom}
\begin{document}
\title{src/backend/global/ExtendedIPC.h}
\author{Dmitry Bormotov}
\maketitle
\begin{abstract}
\end{abstract}
\eject
\tableofcontents
\eject
\section{include}
<<include>>=
#include "global.h"
#include "Integer.h"
#include "Rational.h"

@
\section{Output Operators}
<<Output Operators>>=
inline ostream& operator < ( ostream& ostr, const Integer& n )
{
  ostr << n << ' ';
  return ostr;
}

inline ostream& operator < ( ostream& ostr, const Rational& r )
{
  ostr << r << ' ';
  return ostr;
}

@
\section{Input Operators}
<<Input Operators>>=
inline istream& operator > ( istream& istr, Integer& n )
{
  istr >> n;
  return istr;
}

inline istream& operator > ( istream& istr, Rational& r )
{
  istr >> r;
  return istr;
}

@
\section{license}
<<license>>=
// Copyright (C) 1994-2004 The New York Group Theory Cooperative
// See license/NYGTC for the full notice
@
<<*>>=
// src/backend/global/ExtendedIPC.h
#ifndef _EXTENDED_IPC_H_
#define _EXTENDED_IPC_H_

<<include>>
<<license>>
<<Output Operators>>
<<Input Operators>>
#endif

@
\eject
\begin{thebibliography}{99}
\bibitem{1} nothing
\end{thebibliography}
\end{document}
