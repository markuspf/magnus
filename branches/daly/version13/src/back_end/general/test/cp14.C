// Contents: Computer Project 14
//           Queueing system
//
// Principal Authors: Dmitry Bormotov
//


#include "global.h"
#include "Random.h"

void main( )
{
  UniformRandom r(23152);
  const int numOfCustomers = 20;
  double lambda[numOfCustomers], t[numOfCustomers];
  double x[numOfCustomers], u[numOfCustomers];
  double service[numOfCustomers], departure[numOfCustomers];
  int bound = 21;

  // generate t_n

  double L = 3;
  int count = 0;
  while( count < numOfCustomers ) {
    double time = 1 + r.rand() * (bound-1);
    if( bound * r.rand() <= L / pow(time,4) ) {
      //cout << time << "  " << L * exp(-L * time) << endl;
      t[count++] = time;;
    }
  }

  
  double mean = 0;
  for( int i = 0; i < numOfCustomers; ++i )
    mean += t[i];
  mean /= numOfCustomers;
  cout << endl << "mean = " << mean << endl;
  

  // generate x_n

  L = 4.0 / 3;
  count = 0;
  while( count < numOfCustomers ) {
    double xx = r.rand() * bound;
    if( bound * r.rand() <= L * exp(-L * xx) ) {
      //cout << xx << "  " << L * exp(-L * xx) << endl;
      x[count++] = xx;;
    }
  }

  
  mean = 0;
  for( int i = 0; i < numOfCustomers; ++i )
    mean += x[i];
  mean /= numOfCustomers;
  cout << endl << "mean = " << mean << endl;
  

  // compute lambda_n

  lambda[0] = t[0];
  for( int i = 1; i < numOfCustomers; ++i )
    lambda[i] = t[i] + lambda[i-1];

  // compute u_n

  u[numOfCustomers-1] = 0;
  for( int i = 0; i < numOfCustomers-1; ++i )
    u[i] = x[i] - t[i+1];

  // compute service time and departure time

  double time = lambda[0];
  int customer = 0;

  while( true ) {
    service[customer] = time;
    time += x[customer];
    departure[customer] = time;
    ++customer;
    if( customer >= numOfCustomers ) break;
    if( lambda[customer] > time )
      time = lambda[customer];
  }

  // compute w_n and s_n
  
  double w[numOfCustomers], s[numOfCustomers];
  for( int i = 0; i < numOfCustomers; ++i ) {
    w[i] = service[i] - lambda[i];
    s[i] = departure[i] - lambda[i];
  }

  // compute v_n
 
  double v[numOfCustomers];
  for( int i = 0; i < numOfCustomers; ++i ) {
    v[i] = 0;
    for( int j = 0; j < numOfCustomers; ++j ) 
      if( lambda[j] > service[i] && lambda[j] < departure[i] )
	++v[i];
  }

  // compute g_n
    
  double g[numOfCustomers];
  for( int i = 0; i < numOfCustomers; ++i ) {
    g[i] = 0;
    for( int j = 0; j < numOfCustomers; ++j ) 
      if( lambda[j] < departure[i] && departure[j] > departure[i] )
	++g[i];
  }

  // compute g'_n
    
  double g1[numOfCustomers];
  for( int i = 0; i < numOfCustomers; ++i ) {
    g1[i] = 0;
    for( int j = 0; j < numOfCustomers; ++j ) 
      if( lambda[j] < lambda[i] && departure[j] > lambda[i] )
	++g1[i];
  }

  // compute v'_n+1

  double v1[numOfCustomers];
  v1[0] = 0;
  for( int i = 1; i < numOfCustomers; ++i ) {
    v1[i] = 0;
    for( int j = 0; j < numOfCustomers; ++j ) 
      if( departure[j] > lambda[i-1] && departure[j] < lambda[i] )
	++v1[i];
  }
    
  // print results

  for( int i = 0; i < numOfCustomers; ++i ) {
    cout << "n = " << i+1;
    if( i < 9 ) cout << " ";
    cout << " lambda_n = " << lambda[i] << "  t_n = " << t[i]
	 << "  x_n = " << x[i] << "  u_n = " << u[i] << endl
	 << "       w_n = " << w[i] 
	 << "  s_n = " << s[i] << "  v_n = " << v[i] << "  g_n = " << g[i]
	 << "  g'_n = " << g1[i] << "  v'_n+1 = " << v1[i]
      // << "  service = " << service[i] << "  departure = " << departure[i]
	 << endl << endl << endl; 
  }
}
