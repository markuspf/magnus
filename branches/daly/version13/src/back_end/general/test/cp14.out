n = 1  lambda_n = 1.10253  t_n = 1.10253  x_n = 1.00461  u_n = -0.176822
       w_n = 0  s_n = 1.00461  v_n = 0  g_n = 0  g'_n = 0  v'_n+1 = 0


n = 2  lambda_n = 2.28396  t_n = 1.18144  x_n = 0.43437  u_n = -0.763965
       w_n = 0  s_n = 0.43437  v_n = 0  g_n = 0  g'_n = 0  v'_n+1 = 1


n = 3  lambda_n = 3.4823  t_n = 1.19834  x_n = 0.66225  u_n = -0.687912
       w_n = 0  s_n = 0.66225  v_n = 0  g_n = 0  g'_n = 0  v'_n+1 = 1


n = 4  lambda_n = 4.83246  t_n = 1.35016  x_n = 0.102819  u_n = -2.95204
       w_n = 0  s_n = 0.102819  v_n = 0  g_n = 0  g'_n = 0  v'_n+1 = 1


n = 5  lambda_n = 7.88732  t_n = 3.05485  x_n = 0.11879  u_n = -1.04511
       w_n = 0  s_n = 0.11879  v_n = 0  g_n = 0  g'_n = 0  v'_n+1 = 1


n = 6  lambda_n = 9.05122  t_n = 1.1639  x_n = 0.247076  u_n = -0.806923
       w_n = 0  s_n = 0.247076  v_n = 0  g_n = 0  g'_n = 0  v'_n+1 = 1


n = 7  lambda_n = 10.1052  t_n = 1.054  x_n = 1.86456  u_n = 0.843775
       w_n = 0  s_n = 1.86456  v_n = 1  g_n = 1  g'_n = 0  v'_n+1 = 1


n = 8  lambda_n = 11.126  t_n = 1.02078  x_n = 0.395457  u_n = -0.724482
       w_n = 0.843775  s_n = 1.23923  v_n = 1  g_n = 1  g'_n = 1  v'_n+1 = 0


n = 9  lambda_n = 12.2459  t_n = 1.11994  x_n = 0.358401  u_n = -1.15205
       w_n = 0.119293  s_n = 0.477693  v_n = 0  g_n = 0  g'_n = 1  v'_n+1 = 1


n = 10 lambda_n = 13.7564  t_n = 1.51045  x_n = 1.36467  u_n = 0.0348713
       w_n = 0  s_n = 1.36467  v_n = 1  g_n = 1  g'_n = 0  v'_n+1 = 2


n = 11 lambda_n = 15.0862  t_n = 1.3298  x_n = 0.219845  u_n = -0.845572
       w_n = 0.0348713  s_n = 0.254716  v_n = 0  g_n = 0  g'_n = 1  v'_n+1 = 0


n = 12 lambda_n = 16.1516  t_n = 1.06542  x_n = 2.50126  u_n = 1.10226
       w_n = 0  s_n = 2.50126  v_n = 1  g_n = 1  g'_n = 0  v'_n+1 = 2


n = 13 lambda_n = 17.5506  t_n = 1.399  x_n = 0.17165  u_n = -2.0672
       w_n = 1.10226  s_n = 1.27391  v_n = 0  g_n = 0  g'_n = 1  v'_n+1 = 0


n = 14 lambda_n = 19.7895  t_n = 2.23885  x_n = 0.376513  u_n = -1.62462
       w_n = 0  s_n = 0.376513  v_n = 0  g_n = 0  g'_n = 0  v'_n+1 = 2


n = 15 lambda_n = 21.7906  t_n = 2.00113  x_n = 0.880784  u_n = -3.05836
       w_n = 0  s_n = 0.880784  v_n = 0  g_n = 0  g'_n = 0  v'_n+1 = 1


n = 16 lambda_n = 25.7297  t_n = 3.93914  x_n = 0.160371  u_n = -0.890053
       w_n = 0  s_n = 0.160371  v_n = 0  g_n = 0  g'_n = 0  v'_n+1 = 1


n = 17 lambda_n = 26.7801  t_n = 1.05042  x_n = 1.28135  u_n = -0.278271
       w_n = 0  s_n = 1.28135  v_n = 0  g_n = 0  g'_n = 0  v'_n+1 = 1


n = 18 lambda_n = 28.3398  t_n = 1.55962  x_n = 0.719193  u_n = -1.05701
       w_n = 0  s_n = 0.719193  v_n = 0  g_n = 0  g'_n = 0  v'_n+1 = 1


n = 19 lambda_n = 30.116  t_n = 1.7762  x_n = 0.92833  u_n = -0.132658
       w_n = 0  s_n = 0.92833  v_n = 0  g_n = 0  g'_n = 0  v'_n+1 = 1


n = 20 lambda_n = 31.177  t_n = 1.06099  x_n = 2.10873  u_n = 0
       w_n = 0  s_n = 2.10873  v_n = 0  g_n = 0  g'_n = 0  v'_n+1 = 1
